Add Bang & Olufsen media_player testing

mj23000 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add preliminary testing for the media_player entity for the Bang & Olufsen integration.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

mj23000 committed:
Initial work

mj23000 committed:
Add turn_off testing
Add API call checking

mj23000 committed:
Add volume testing

mj23000 committed:
Clean up state assertion

mj23000 committed:
Further tests added

mj23000 committed:
Fix deezer test

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_media_player_testing

mj23000 committed:
Add remaining initial tests

mj23000 committed:
Improve tests

mj23000 committed:
Remove unnecessary mock client patch

mj23000 committed:
Improve exception assertion

davet2001 reviewed (commented):
Thanks @mj23000!

Minor comments:
- There are lots of comment lines, following a very similar pattern each time. To me this is bordering on more than necessary (eg 'setup entity', 'send a service call') on every test. Perhaps comment the first one in detail, then be more minimal on the rest?
- There are some very similar tests. Have you considered `@pytest.mark.parametrize`? This can help with maintenance.

But overall looks good! Thanks for taking the time to write the tests.

mj23000 committed:
Remove repetitive and redundant comments

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_media_player_testing

mj23000 committed:
Update tests for new error messages

mj23000 commented:
> Thanks @mj23000!
> 
> Minor comments:
> 
>     * There are lots of comment lines, following a very similar pattern each time. To me this is bordering on more than necessary (eg 'setup entity', 'send a service call') on every test. Perhaps comment the first one in detail, then be more minimal on the rest?
> 
>     * There are some very similar tests. Have you considered `@pytest.mark.parametrize`? This can help with maintenance.
> 
> 
> But overall looks good! Thanks for taking the time to write the tests.

Thank you for the feedback. I've removed the redundant comments and updated the tests to use the new HomeAssistantError raised exceptions instead of _LOGGER.error messages. 

I am however having issues getting `test_async_update_playback_error` to trigger pytest.raises, probably because I'm using `async_dispatcher_send`. Do you have any idea how I could fix this? It's currently commented out while I try to figure this issue out.

mj23000 committed:
Parameterize various tests
Fix broken source test(?)

davet2001 commented:
@mj23000 I think the Exceptions are being raised, they are just being swallowed by `async_dispatcher_send_internal` [here](https://github.com/home-assistant/core/blob/e764afecac101f24154f77394c81607647d45779/homeassistant/helpers/dispatcher.py#L245-L249). Which captures broad exceptions for callbacks and turns them into log messages.

So I think you need to look at the logs for evidence of the Exception if you want to see that it happened.

mj23000 committed:
Fix test_async_update_playback_error by using logger

mj23000 commented:
> @mj23000 I think the Exceptions are being raised, they are just being swallowed by `async_dispatcher_send_internal` [here](https://github.com/home-assistant/core/blob/e764afecac101f24154f77394c81607647d45779/homeassistant/helpers/dispatcher.py#L245-L249). Which captures broad exceptions for callbacks and turns them into log messages.
> 
> So I think you need to look at the logs for evidence of the Exception if you want to see that it happened.

Good idea, thanks. I've re-added the test with logger checking.

mj23000 committed:
Merge branch 'home-assistant:dev' into bang_olufsen_add_media_player_testing

davet2001 reviewed (changes requested):
The tests are almost complete coverage. Thanks!
It feels like getting 100% is achievable here, only 6 items missing.
```
pytest tests/components/bang_olufsen/ --cov=homeassistant.components.bang_olufsen --cov-report term-missing -vv
```
...will tell you what is missing.

More specifically, I could not get the test coverage to increase by setting `source.id` to one of the values in the search list. When I did it caused an error. All the more reason to include tests for those lines.

mj23000 commented:
> The tests are almost complete coverage. Thanks! It feels like getting 100% is achievable here, only 6 items missing.
> 
> ```
> pytest tests/components/bang_olufsen/ --cov=homeassistant.components.bang_olufsen --cov-report term-missing -vv
> ```
> 
> ...will tell you what is missing.
> 
> More specifically, I could not get the test coverage to increase by setting `source.id` to one of the values in the search list. When I did it caused an error. All the more reason to include tests for those lines.

Yes most of the untested source lines are caused by the currently broken source checks (Which should be fixed here https://github.com/home-assistant/core/pull/116052). I'll take a look at `async_browse_media` for now.

mj23000 committed:
Add async_browse_media testing

davet2001 commented:
@mj23000 Please set the PR to 'ready for review' when you are ready.

mj23000 committed:
Merge branch 'dev' into bang_olufsen_add_media_player_testing

davet2001 commented:
Hi @mj23000 looking good, thanks for this!

Happy to approve. Even better if you could get the last few lines of test coverage:

https://app.codecov.io/gh/home-assistant/core/pull/120580/blob/homeassistant/components/bang_olufsen/media_player.py

mj23000 committed:
Merge branch 'home-assistant:dev' into bang_olufsen_add_media_player_testing

mj23000 committed:
Add more source testing

mj23000 committed:
Merge branch 'home-assistant:dev' into bang_olufsen_add_media_player_testing

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+    await hass.services.async_call(
+        "media_player",
+        "turn_off",
+        {ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID},
```
```suggestion
        {ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID},
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+        {
+            ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID,
+            ATTR_MEDIA_VOLUME_LEVEL: TEST_VOLUME_HOME_ASSISTANT_FORMAT,
+        },
```
```suggestion
        },
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+        {
+            ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID,
+            ATTR_MEDIA_VOLUME_MUTED: TEST_VOLUME_HOME_ASSISTANT_FORMAT,
+        },
```
```suggestion
        },
        blocking=True,
```

mj23000 committed:
Merge branch 'bang_olufsen_add_media_player_testing' of https://github.com/mj23000/core into bang_olufsen_add_media_player_testing

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+    await hass.services.async_call(
+        "media_player",
+        "media_play_pause",
+        {ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID},
```
```suggestion
        {ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID},
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+    await hass.services.async_call(
+        "media_player",
+        "media_next_track",
+        {ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID},
```
```suggestion
        {ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID},
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+    await hass.services.async_call(
+        "media_player",
+        "media_previous_track",
+        {ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID},
```
```suggestion
        {ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID},
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+    await hass.services.async_call(
+        "media_player",
+        "clear_playlist",
+        {ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID},
```
```suggestion
        {ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID},
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+            ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID,
+            ATTR_MEDIA_CONTENT_ID: "media-source://media_source/local/doorbell.mp3",
+            ATTR_MEDIA_CONTENT_TYPE: "audio/mpeg",
+        },
```
```suggestion
        },
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+            ATTR_MEDIA_CONTENT_TYPE: "music",
+            ATTR_MEDIA_ANNOUNCE: True,
+            ATTR_MEDIA_EXTRA: {"overlay_absolute_volume": 60},
+        },
```
```suggestion
        },
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+                    "overlay_offset_volume": 20,
+                    "overlay_tts_language": "da-dk",
+                },
+            },
```
```suggestion
            },
            blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+            ATTR_MEDIA_CONTENT_TYPE: "overlay_tts",
+            ATTR_MEDIA_ANNOUNCE: True,
+            ATTR_MEDIA_EXTRA: {"overlay_offset_volume": 20},
+        },
```
```suggestion
        },
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+            ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID,
+            ATTR_MEDIA_CONTENT_ID: "media-source://media_source/local/doorbell.mp3",
+            ATTR_MEDIA_CONTENT_TYPE: "provider",
+        },
```
```suggestion
        },
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+            ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID,
+            ATTR_MEDIA_CONTENT_ID: "1234567890123456",
+            ATTR_MEDIA_CONTENT_TYPE: "radio",
+        },
```
```suggestion
        },
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+            ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID,
+            ATTR_MEDIA_CONTENT_ID: "1",
+            ATTR_MEDIA_CONTENT_TYPE: "favourite",
+        },
```
```suggestion
        },
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+            ATTR_MEDIA_CONTENT_ID: "flow",
+            ATTR_MEDIA_CONTENT_TYPE: "deezer",
+            ATTR_MEDIA_EXTRA: {"id": "123"},
+        },
```
```suggestion
        },
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+            ATTR_MEDIA_CONTENT_ID: "playlist:1234567890",
+            ATTR_MEDIA_CONTENT_TYPE: "deezer",
+            ATTR_MEDIA_EXTRA: {"start_from": 123},
+        },
```
```suggestion
        },
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+            ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID,
+            ATTR_MEDIA_CONTENT_ID: "1234567890",
+            ATTR_MEDIA_CONTENT_TYPE: "deezer",
+        },
```
```suggestion
        },
        blocking=True,
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,984 @@
+    await async_setup_component(hass, "media_source", {"media_source": {}})
+
+    client = await hass_ws_client()
+    await client.send_json(
+        {
+            "id": 1,
```
```suggestion
    await client.send_json_auto_id(
        {
```

frenck commented on tests/components/bang_olufsen/test_media_player.py:
```diff
@@ -0,0 +1,1050 @@
+    await hass.services.async_call(
+        "media_player",
+        "media_stop",
+        {ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID},
```
```suggestion
        {ATTR_ENTITY_ID: TEST_MEDIA_PLAYER_ENTITY_ID},
        blocking=True,
```

mj23000 committed:
Add final source test

mj23000 committed:
Make service calls blocking
Add audio ID to websocket client call

mj23000 commented:
Thank you both for the feedback. Code coverage is now at 100% for the media_player entity, so the tests should be done for now.

frenck reviewed (approved):
Thanks, @mj23000 👍 

../Frenck
